### PR TITLE
Upgrades overhaul part2: GrantCondition and Carryall fixes.

### PR DESCRIPTION
**Pull request by pchote**
This implements the `GrantCondition` trait that was outlined (as `ConditionCombiner`) in #12380.

This also ports the carryall traits over to the new conditions code, and works around the glitches with tick tanks and mobile sensor arrays.

Depends on #12381.

### CODE REVIEW

**Comment by Phrohdoh** [comment]
Why not name it `ConditionCombiner` which is more explanatory?

**Comment by pchote** [comment]
Because it doesn't have any special logic to combine conditions.  All it does is grant a condition, and like other traits, this can be enabled or disabled depending on other conditions.

**Comment by Phrohdoh** [comment]
The purpose is combining (one or more) conditions into a single one.
By your logic every condition trait ought to be `GrantCondition`.

I think `ConditionCombiner` is more helpful (imagine someone coming to IRC and asking "how do I combine conditions?" the answer is easy to explain and actually in their question).

We ought to err on the side of too wordy and helpful rather than being vague, IMO.

**Comment by pchote** [comment]
The only use case we have for this trait right now (included in this PR) is to unconditionally grant a condition when an actor has been created.

**Review by reaperrr — APPROVED** [review]

**Comment by reaperrr** [comment]
Needs rebase.

**Comment by pchote** [comment]
Rebased.

File: OpenRA.Mods.Common/Traits/Carryable.cs [hunk 1]
```diff
@@ -15,38 +15,41 @@
...
 {
 	[Desc("Can be carried by actors with the `Carryall` trait.")]
-	public class CarryableInfo : ITraitInfo
+	public class CarryableInfo : UpgradableTraitInfo
 	{
 		[UpgradeGrantedReference]
-		[Desc("The upgrades to grant to self while waiting or being carried.")]
-		public readonly string[] CarryableUpgrades = { };
+		[Desc("The condition to grant to self while a carryall has been reserved.")]
+		public readonly string ReservedCondition = "carryall-reserved";
```
Comment (pchote) [inline OpenRA.Mods.Common/Traits/Carryable.cs]: Does anyone have an opinion about granting these automatically (where it isn't obvious to modders where they are granted) vs defaulting to nothing and requiring the mods to define them (redundancy and trickier upgrade rules).

I'm leaning towards the second option, so adding the fixup requested tag until I have time to fix it.

Comment (GraionDilach) [inline OpenRA.Mods.Common/Traits/Carryable.cs]: Yes, please go for the requiring-for-mods-to-define approach.

**Comment by pchote** [comment]
This pr is blocking a bunch more work, so can we please get a second :+1: and merge?

**Comment by GraionDilach** [comment]
Sure, I was just testing this PR while you were writing that tbh. :+1:

**Comment by reaperrr** [comment]
[changelog](https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/455d8a5639323686751636c1f7e6dd2de41d4739...8c23a5607eb89d0955a80c852d2ff7ee07c5c56c)
